### PR TITLE
Add simulation status reporting

### PR DESCRIPTION
**Pull request by Jim3Things**

### CODE REVIEW

File: internal/services/frontend/httpErrors.go [hunk 1]
```diff
@@ -11,6 +11,7 @@ import (
 	"strconv"
 
 	"github.com/Jim3Things/CloudChamber/internal/tracing"
+	err2 "github.com/Jim3Things/CloudChamber/pkg/errors"
```
Comment (MikeCJohnsonUk) [inline internal/services/frontend/httpErrors.go:14]: I found myself reacting to the "err2", the '2' part in particular.

Would perhaps just "e" suffice here?

File: internal/services/frontend/session_manager.go [hunk 2]
```diff
@@ -85,6 +106,24 @@ func removeSession(session *sessions.Session) {
...
+// use by the caller.
+func removeSessionById(id int64) (sessionState, bool) {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	purgeStaleSessions()
+
+	if entry, ok2 := activeSessions[id]; ok2 {
+		delete(activeSessions, id)
+		delete(timeouts, entry.timeout)
```
Comment (MikeCJohnsonUk) [inline internal/services/frontend/session_manager.go:120]: Should these two lines be commoned up with those in purgeStaleSessions() into a deleteSession() func to ensure the same delete actions are always taken for a session removal?

File: internal/services/frontend/session_manager.go [hunk 3]
```diff
@@ -116,6 +155,20 @@ func getSession(session *sessions.Session) (sessionState, bool) {
...
+
+	purgeStaleSessions()
+
+	entry, ok := activeSessions[id]
+	return entry, ok
+}
+
 // purgeStaleSessions removes stale sessions from the active session list.  It
 // assumes that the number of sessions is not so large that the active session
 // list cannot be scanned.
```
Comment (MikeCJohnsonUk) [inline internal/services/frontend/session_manager.go:174]: Do we have anything anywhere which limits the number of active sessions? Should we? Just to stop the runaways.

File: internal/services/frontend/simulation_test.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+	request = httptest.NewRequest("GET", status.Sessions[0].Uri, nil)
+	request.Header.Set("Content-Type", "application/json")
+
+	response = ts.doHTTP(request, response.Cookies())
+	require.Equal(http.StatusForbidden, response.StatusCode)
+}
+
+func TestSimulationTestSuite(t *testing.T) {
+	suite.Run(t, new(SimulationTestSuite))
+}
```
Comment (MikeCJohnsonUk) [inline internal/services/frontend/simulation_test.go:144]: Do we need anything to cover non-admin sessions. non-admin calls and multiple active sessions?

File: internal/services/frontend/httpErrors.go (hunk 1 above)
Comment (Jim3Things) [reply internal/services/frontend/httpErrors.go:14]: I'll try that.

File: internal/services/frontend/session_manager.go (hunk 3 above)
Comment (Jim3Things) [reply internal/services/frontend/session_manager.go:174]: We don't.  I'll add an issue.

File: internal/services/frontend/simulation_test.go (hunk 4 above)
Comment (Jim3Things) [reply internal/services/frontend/simulation_test.go:144]: Yeah, should do.  Forgot that when doing the tests.

File: internal/services/frontend/session_manager.go (hunk 2 above)
Comment (Jim3Things) [reply internal/services/frontend/session_manager.go:120]: Hmm.  Yeah, not much of a gain, but yeah, it does replace some repeat code.

I also realize that there's a bug here - we do the wrong thing on timeout collisions.  I'm not sure we'd see such a collision, but I should fix that anyway.

Comment (MikeCJohnsonUk) [reply internal/services/frontend/session_manager.go:120]: I wasn't thinking about the savings per-se, more that we have a consistent set of operations. I've seen other project run afoul of that.

But the savings, are nice too.

File: internal/services/frontend/session_manager.go [hunk 5]
```diff
@@ -214,58 +199,229 @@ func dumpSessionState(session *sessions.Session) string {
...
 		}
 	}
 
-	user, _, err := dbUsers.Read(ctx, entry.name)
-	return user, err
+	st.lastID++
+	entry.timeout = time.Now().Add(st.maxInactive)
+	st.addToTables(st.lastID, entry)
+
+	return st.lastID, nil
```
Comment (MikeCJohnsonUk) [inline internal/services/frontend/session_manager.go:273]: Possibly over picky, but when exactly does the deferred unlock happen wrt the acquisition of the current value of st.lastID? 

I think this is ok, but just wanted to check.

File: internal/services/frontend/session_manager.go [hunk 6]
```diff
@@ -214,58 +199,229 @@ func dumpSessionState(session *sessions.Session) string {
...
+// Note that expired entries are first removed from the sessionTable prior to
+// processing this request.
+func (st *sessionTable) touch(id int64) (sessionState, bool) {
+	st.m.Lock()
+	defer st.m.Unlock()
+
+	st.purge()
+
+	if entry, ok := st.known[id]; ok {
+		st.deleteFromTimeoutList(id, entry.timeout)
```
Comment (MikeCJohnsonUk) [inline internal/services/frontend/session_manager.go:321]: being picky again: removeFromTimeoutList() ?

File: internal/services/frontend/session_manager.go (hunk 5 above)
Comment (Jim3Things) [reply internal/services/frontend/session_manager.go:273]: According to go, defer executes after the return, so we should be good.

File: internal/services/frontend/session_manager.go (hunk 6 above)
Comment (Jim3Things) [reply internal/services/frontend/session_manager.go]: yeah, that's pretty picky :).  Done.

**Review by MikeCJohnsonUk — APPROVED** [review]

**Pull request merged**